components/browser/engine-servo/src/main/java/mozilla/components/browser/engine/servo/ServoEngineSession.kt

```diff
@@ -0,0 +1,150 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.engine.servo
+
+import android.graphics.Bitmap
+import android.net.Uri
+import mozilla.components.concept.engine.DefaultSettings
+import mozilla.components.concept.engine.EngineSession
+import mozilla.components.concept.engine.Settings
+import org.mozilla.servoview.Servo
+import org.mozilla.servoview.ServoView
+import java.lang.ref.WeakReference
+
+/**
+ * Servo-based EngineSession implementation.
+ */
+@Suppress("TooManyFunctions")
+class ServoEngineSession(
+    private val defaultSettings: Settings? = null
+) : EngineSession() {
+    override val settings: Settings
+        get() = defaultSettings ?: DefaultSettings()
+
+    private var urlToLoad: String? = null
+    private var viewReference: WeakReference<ServoEngineView>? = null
+    private val view: ServoView?
+        get() = viewReference?.get()?.servoView
+
+    @Suppress("TooGenericExceptionCaught")
+    internal fun attachView(view: ServoEngineView) {
+        viewReference = WeakReference(view)
+
+        view.servoView.setClient(client)
+        view.servoView.requestFocus()
+
+        urlToLoad?.let {
+            view.servoView.loadUri(Uri.parse(it))
+        }
+    }
+
+    private val client = object : Servo.Client {
+        override fun onLoadStarted() {
+            notifyObservers {
+                onLoadingStateChange(true)
+            }
+        }
+
+        override fun onRedrawing(redrawing: Boolean) = Unit
+
+        override fun onTitleChanged(title: String) {
+            notifyObservers {
+                onTitleChange(title)
+            }
+        }
+
+        override fun onLoadEnded() {
+            notifyObservers {
+                onLoadingStateChange(false)
+            }
+        }
+
+        override fun onUrlChanged(url: String) {
+            notifyObservers {
+                onLocationChange(url)
+            }
+        }
+
+        override fun onHistoryChanged(canGoBack: Boolean, canGoForward: Boolean) {
+            notifyObservers {
+                onNavigationStateChange(canGoBack, canGoForward)
+            }
+        }
+    }
+
+    override fun loadUrl(url: String) {
+        val view = view
```

Comment:
This seems like an unneeded assignment?

Reply:
The `view` property is nullable and with the local reference I circumvent the issue that it may change after the null check. Usually we do this with `view?.let {}` but I need an `else` branch here as well.

Reply:
heh, yes, that's how we usually end up with `view?.let {} ?: (run) {}`. Still on the fence what's easier to read...